From ee7dcff908c8d64be4c789a25e90b80c0298273f Mon Sep 17 00:00:00 2001
From: Vishal Sagar <vishal.sagar@xilinx.com>
Date: Fri, 21 Sep 2018 15:53:48 +0530
Subject: [PATCH 0422/1852] v4l: xilinx-demosaic: Add check for max width and
 height

commit da26d7ecbf3de892ada21e2efadb0fa3d2bfae56 from
https://github.com/Xilinx/linux-xlnx.git

Patch adds support to get maximum width and height from device tree
property. If the parameters are not present then the IP sets the
maximum size of 8192x4320. It then uses this value to clamp width
and height instead of fixed default value while setting format.

It also corrects the minimum width and height parameters.

Signed-off-by: Vishal Sagar <vishal.sagar@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 .../media/platform/xilinx/xilinx-demosaic.c   | 34 +++++++++++++++----
 1 file changed, 28 insertions(+), 6 deletions(-)

diff --git a/drivers/media/platform/xilinx/xilinx-demosaic.c b/drivers/media/platform/xilinx/xilinx-demosaic.c
index 505fc14dcd79..ff7a39fdb962 100644
--- a/drivers/media/platform/xilinx/xilinx-demosaic.c
+++ b/drivers/media/platform/xilinx/xilinx-demosaic.c
@@ -26,11 +26,11 @@
 #define XDEMOSAIC_HEIGHT			(0x18)
 #define XDEMOSAIC_INPUT_BAYER_FORMAT		(0x28)
 
-#define XDEMOSAIC_MIN_HEIGHT	(32)
-#define XDEMOSAIC_MAX_HEIGHT	(2160)
+#define XDEMOSAIC_MIN_HEIGHT	(64)
+#define XDEMOSAIC_MAX_HEIGHT	(4320)
 #define XDEMOSAIC_DEF_HEIGHT	(720)
-#define XDEMOSAIC_MIN_WIDTH	(32)
-#define XDEMOSAIC_MAX_WIDTH	(3840)
+#define XDEMOSAIC_MIN_WIDTH	(64)
+#define XDEMOSAIC_MAX_WIDTH	(8192)
 #define XDEMOSAIC_DEF_WIDTH	(1280)
 
 #define XDEMOSAIC_RESET_DEASSERT	(0)
@@ -54,6 +54,8 @@ struct xdmsc_dev {
 
 	enum xdmsc_bayer_format bayer_fmt;
 	struct gpio_desc *rst_gpio;
+	u32 max_width;
+	u32 max_height;
 };
 
 static inline u32 xdmsc_read(struct xdmsc_dev *xdmsc, u32 reg)
@@ -171,9 +173,9 @@ static int xdmsc_set_format(struct v4l2_subdev *subdev,
 	*__format = fmt->format;
 
 	__format->width = clamp_t(unsigned int, fmt->format.width,
-				XDEMOSAIC_MIN_WIDTH, XDEMOSAIC_MAX_WIDTH);
+				  XDEMOSAIC_MIN_WIDTH, xdmsc->max_width);
 	__format->height = clamp_t(unsigned int, fmt->format.height,
-				XDEMOSAIC_MIN_HEIGHT, XDEMOSAIC_MAX_HEIGHT);
+				   XDEMOSAIC_MIN_HEIGHT, xdmsc->max_height);
 
 	if (fmt->pad == XVIP_PAD_SOURCE) {
 		if (__format->code != MEDIA_BUS_FMT_RBG888_1X24) {
@@ -244,6 +246,26 @@ static int xdmsc_parse_of(struct xdmsc_dev *xdmsc)
 	u32 port_id = 0;
 	int rval;
 
+	rval = of_property_read_u32(node, "xlnx,max-height",
+				    &xdmsc->max_height);
+	if (rval < 0) {
+		xdmsc->max_height = XDEMOSAIC_MAX_HEIGHT;
+	} else if (xdmsc->max_height > XDEMOSAIC_MAX_HEIGHT ||
+		 xdmsc->max_height < XDEMOSAIC_MIN_HEIGHT) {
+		dev_err(dev, "Invalid height in dt");
+		return -EINVAL;
+	}
+
+	rval = of_property_read_u32(node, "xlnx,max-width",
+				    &xdmsc->max_width);
+	if (rval < 0) {
+		xdmsc->max_width = XDEMOSAIC_MAX_WIDTH;
+	} else if (xdmsc->max_width > XDEMOSAIC_MAX_WIDTH ||
+		 xdmsc->max_width < XDEMOSAIC_MIN_WIDTH) {
+		dev_err(dev, "Invalid width in dt");
+		return -EINVAL;
+	}
+
 	ports = of_get_child_by_name(node, "ports");
 	if (!ports)
 		ports = node;
-- 
2.31.1

